### ISSUE TITLE
Add support for passing IBM Adjunct Processor (AP) devices over VFIO

### ISSUE THREAD

**Issue by Jakob-Naucke** [kata-containers/kata-containers#491]
**Which feature do you think can be improved?**

IBM offers a range of hardware security modules (HSMs), the [Crypto Express Cards](https://www.ibm.com/security/cryptocards), which connect over PCIe. However, on the IBM Z mainframe platform (s390x), the cards are actually exposed over another bus, the Adjunct Processor (AP) bus. This bus has also [received support for VFIO](https://www.kernel.org/doc/html/latest/s390/vfio-ap.html), using mediated devices.

Passing PCI(e) devices such as GPUs over VFIO is already supported in Kata. Kata's functionality could be enhanced by supporting the pass-through of such AP devices. This would make the HSM functionalities available inside Kata containers, such as hardware accelerated cryptographic algorithms, key management, random and prime number generation, and tamper resistance.

**How can it be improved?**

I have written patches to support this. One is in [GoVMM](https://github.com/intel/govmm), where I added functionality for hotplugging AP devices besides PCI(e) devices. I have also written an s390x fragment for the kernel options necessary for VFIO-AP in `tools/packaging`. Finally, I implemented logic in `src/runtime/virtcontainers` the logic to recognise that the device to be hotplugged is an AP device rather than a PCI device and that the respective functionality in GoVMM be used.

VFIO-AP pass-through to a Kata container then works by making a card and domain available for VFIO and assigning them to a mediated device as per the [kernel documentation](https://www.kernel.org/doc/html/latest/s390/vfio-ap.html), determining the IOMMU group for this mediated device similar to [Kata's guide on doing this with an Intel i915 GPU](https://github.com/kata-containers/kata-containers/blob/2.0-dev/docs/use-cases/Intel-GPU-passthrough-and-Kata.md), and running something like

```sh
$ ctr run --runtime io.containerd.run.kata.v2 -t --rm --device /dev/vfio/0 docker.io/library/ubuntu:latest foo bash
$ lszcrypt # requires s390-tools to be installed
CARD.DOMAIN TYPE  MODE        STATUS  REQUESTS
----------------------------------------------
09          CEX6C CCA-Coproc  online         1
09.003d     CEX6C CCA-Coproc  online         1
```

**Additional Information**

I intend to raise a PR for my GoVMM changes first. I would then re-vendor this project for the GoVMM changes.
In parallel, I would also need a forward-port of s390x fragments in packaging, i.e. PRs [#1042](https://github.com/kata-containers/packaging/pull/1042) and [#1046](https://github.com/kata-containers/packaging/pull/1046). Should that be done as a re-vendoring as well? Given that those re-vendorings are accepted, I would raise PRs for packaging and virtcontainers.

Regarding documentation: The required steps are really just a combination of Linux' VFIO-AP documentation and Kata's GPU passthrough documentation. Does this warrant a dedicated documentation use case?

The work on this was backed and reviewed by @alicefr.

**Before raising this feature request**

I have looked at the [limitations document](https://github.com/kata-containers/documentation/blob/master/Limitations.md).

**Comment by alicefr** [comment]
@Jakob-Naucke for the porting of #1042 and #1046 you might need to create a separate issue.

**Comment by Jakob-Naucke** [comment]
I now noticed that [packaging's #1042](https://github.com/kata-containers/packaging/pull/1042) was already there.

**Comment by Jakob-Naucke** [comment]
@devimc I feel documentation hasn't been addressed enough yet.
I apologise for reopening this here and for not further investigating the missing WIP label that I had wanted to put on my PR because of this (but don't worry, the code is fine).
I had previously written about [Kata's guide on Intel iGPU pass-through](https://github.com/kata-containers/kata-containers/blob/2.0-dev/docs/use-cases/Intel-GPU-passthrough-and-Kata.md) already kind of describing what to do, while having to follow the [kernel documentation](https://www.kernel.org/doc/html/latest/s390/vfio-ap.html) for preparation. I felt that mentioning this nowhere in the documentation would be a bit too little, and writing a dedicated guide when it's so similar would be a bit redundant.
More recently, I noticed that the [virtcontainers README](https://github.com/kata-containers/kata-containers/tree/2.0-dev/src/runtime/virtcontainers#how-to-pass-a-device-using-vfio-passthrough) also mentions VFIO, while being somewhat PCI specific.
WDYT? Is documentation fine as is? Should there be a side note that is is also possible with AP somewhere (where?)? Should there be a full guide?

**Comment by devimc** [comment]
> I had previously written about Kata's guide on Intel iGPU pass-through already kind of describing what to do, while having to follow the kernel documentation for preparation. I felt that mentioning this nowhere in the documentation would be a bit too little, and writing a dedicated guide when it's so similar would be a bit redundant.

you should update it either way to reflect your changes, you can add a section or a note for s390x when needed, btw do Intel GPUs are compatible with s390x AP ? what about nvidia?

> More recently, I noticed that the virtcontainers README also mentions VFIO, while being somewhat PCI specific.
WDYT? Is documentation fine as is? Should there be a side note that is is also possible with AP somewhere (where?)? Should there be a full guide?

I'd rename that section and add a new one

How to pass a device using VFIO-passthrough -> How to pass a device using VFIO-PCI-passthrough

How to pass a device using VFIO-AP-passthrough (s390x)

**Comment by alicefr** [comment]
> btw do Intel GPUs are compatible with s390x AP ? what about nvidia?

@devimc there is not GPU support on s390x either for Intel or Nvidia. Only Power support GPUs :)

**Comment by alicefr** [comment]
vfio ap is used for the Crypto Express Card: https://www.ibm.com/security/cryptocards

**Comment by devimc** [comment]
> @devimc there is not GPU support on s390x either for Intel or Nvidia. Only Power support GPUs :)

thanks @alicefr , @Jakob-Naucke  we'll need a new guide for ` Power support GPUs` , what do you think ?

**Comment by alicefr** [comment]
@devimc you need to ping the Power maintainer for Kata (not sure who is after @nitkon). But it makes of course sense.
@Jakob-Naucke only has access to Z systems. 
BTW the vfio ap bus on s390x is used for cryptographic devices.

**Comment by Jakob-Naucke** [comment]
@devimc 
> I'd rename that section and add a new one

Will do.
> we'll need a new guide for `Power support GPUs` , what do you think ?

My general assumption would be that passing a GPU to a Kata container on IBM POWER works just like it does on x86, i.e. normal PCI VFIO pass-through. Thus, no new guide would be required. However, I don't really do POWER – possibly @bpradipt could help you out.

**Comment by devimc** [comment]
might be @bpradipt can help us here,

**Comment by devimc** [comment]
> My general assumption would be that passing a GPU to a Kata container on IBM POWER works just like it does on x86, i.e. normal PCI VFIO pass-through. Thus, no new guide would be required. However, I don't really do POWER – possibly @bpradipt could help you out.

fair enough, since @bpradipt is the only one with access to POWER HW, it's up to him document it or not :smile:
